grind_75/week_1/medium/number-of-islands.py

```diff
@@ -0,0 +1,48 @@
+# https://leetcode.com/problems/number-of-islands/submissions/
+
+class UnionFind:
+    def __init__(self, count):
+        self.count = count
+        self.roots = [i for i in range(count)]
+
+    def find(self, x):
+        if self.roots[x] == x:
+            return x
+        return self.find(self.roots[x])
+
+    def pair(self, x, y):
+        root_x = self.find(x)
+        root_y = self.find(y)
+        print(x,y, root_x, root_y)
+        if root_x == root_y:
+            return
+        if root_x < root_y:
+            self.roots[root_y] = root_x
+        else:
+            self.roots[root_x] = root_y
+        self.count -= 1
```

Comment:
Grandparent size reduction / Weight compression is very important to make the complexity -> O(N * Ack(N))

where Ack(N) is iterated logarithm. https://en.wikipedia.org/wiki/Iterated_logarithm